cmd/meroxa/root/apps/deploy.go

```diff
@@ -102,24 +118,53 @@ func (d *Deploy) Logger(logger log.Logger) {
 	d.logger = logger
 }
 
+func (d *Deploy) createApplication(ctx context.Context) error {
+	appName, err := turbineCLI.GetAppNameFromAppJSON(d.path)
+	if err != nil {
+		return err
+	}
+
+	input := meroxa.CreateApplicationInput{
+		Name:     appName,
+		Language: d.lang,
+	}
+	d.logger.Infof(ctx, "Creating application %q with language %q...", input.Name, d.lang)
+
+	res, err := d.client.CreateApplication(ctx, &input)
+	if err != nil {
+		return err
+	}
+
+	d.logger.Infof(ctx, "Application %q successfully created!", res.Name)
+	d.logger.JSON(ctx, res)
+	return nil
+}
+
 func (d *Deploy) Execute(ctx context.Context) error {
 	err := d.checkRequiredEnvVars()
 	if err != nil {
 		return err
 	}
 
 	d.path = turbineCLI.GetPath(d.flags.Path)
-	lang, err := turbineCLI.GetLangFromAppJSON(d.path)
+	d.lang, err = turbineCLI.GetLangFromAppJSON(d.path)
 	if err != nil {
 		return err
 	}
 
-	switch lang {
-	case "go", GoLang:
-		return d.goDeploy.DeployGoApp(ctx, d.path, d.logger)
+	switch d.lang {
+	case GoLang:
+		err = d.goDeploy.DeployGoApp(ctx, d.path, d.logger)
 	case "js", JavaScript, NodeJs:
-		return turbineCLI.DeployJSApp(ctx, d.path, d.logger)
+		err = turbineCLI.DeployJSApp(ctx, d.path, d.logger)
 	default:
-		return fmt.Errorf("language %q not supported. Currently, we support \"javascript\" and \"go\"", lang)
+		return fmt.Errorf("language %q not supported. %s", d.lang, LanguageNotSupportedError)
+	}
+
+	if err != nil {
+		return err
 	}
+
+	// Built was successful, we're ready to create the application
```

Comment:
I agree with @janelletavares that the user experience could be improved by having the app name validation (which is almost immediate) fail before the source upload and function image build have finished completely (which would take some time)

I can imagine though that the creation of connectors could be postponed to a later time point than the function build. The same would apply to creating the function objects which depend on the function image build to finish beforehand

I don't want to block this PR from moving forward, just mentioning this already, in case there's interest on iterating on it in the future

Reply:
> I agree with @janelletavares that the user experience could be improved by having the app name validation (which is almost immediate) fail before the source upload and function image build have finished completely (which would take some time)

We documented we'd be creating the app only if build was successful (we'll be doing a long polling and blocking the next steps until building is successful) so we don't need to worry about orphan apps and then let users do another re-deploy in an attempt to build again (we don't have app updates available).

I understand what you two are mentioning regarding the UX, and I'll give this another thought however what I'm trying to accomplish here is to leave the step that less likely will fail to the end. In other words, trying to build several times and leave different building steps hanging is completely fine, however having an application created (and not deleted) if something after failed not so much (we're creating turbine apps after all).

Happy to discuss different corner cases and possible scenarios (with its UX improvements).

Reply:
interesting about doing the steps most likely to fail first. thanks for the clarification.

Reply:
That sounds good and thank you for the outline!